worker/provisioner/provisioner_task.go

```diff
@@ -689,106 +714,357 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+	}
+
+	instanceCfg, err := task.constructInstanceConfig(machine, task.auth, pInfo)
+	if err != nil {
+		return nil, environs.StartInstanceParams{}, task.setErrorStatus("creating instance config for machine %q: %v", machine, err)
+	}
+
+	assocProvInfoAndMachCfg(pInfo, instanceCfg)
+
+	var arch string
+	if pInfo.Constraints.Arch != nil {
+		arch = *pInfo.Constraints.Arch
+	}
+
+	v, err := machine.ModelAgentVersion()
+	if err != nil {
+		return nil, environs.StartInstanceParams{}, errors.Trace(err)
+	}
+	possibleTools, err := task.toolsFinder.FindTools(
+		*v,
+		pInfo.Series,
+		arch,
+	)
+	if err != nil {
+		return nil, environs.StartInstanceParams{}, task.setErrorStatus("cannot find tools for machine %q: %v", machine, err)
+	}
+
+	// ToDo ProvisionerParallelization 2017-10-03
+	// Fix StartInstanceParams struct to carry the availability zone,
+	// overwrite the placement as done currently.
+	// DistributionGroups() can be removed.
+	startInstanceParams, err := constructStartInstanceParams(
+		task.controllerUUID,
+		machine,
+		instanceCfg,
+		pInfo,
+		possibleTools,
+		zone,
+	)
+	if err != nil {
+		return nil, environs.StartInstanceParams{}, task.setErrorStatus("cannot construct params for machine %q: %v", machine, err)
+	}
+
+	return pInfo, startInstanceParams, nil
 }
 
 func (task *provisionerTask) startMachine(
 	machine *apiprovisioner.Machine,
 	provisioningInfo *params.ProvisioningInfo,
 	startInstanceParams environs.StartInstanceParams,
 ) error {
-	var result *environs.StartInstanceResult
 	// TODO (jam): 2017-01-19 Should we be setting this earlier in the cycle?
 	if err := machine.SetInstanceStatus(status.Provisioning, "starting", nil); err != nil {
 		logger.Errorf("%v", err)
 	}
-	for attemptsLeft := task.retryStartInstanceStrategy.retryCount; attemptsLeft >= 0; attemptsLeft-- {
+
+	// ToDo ProvisionerParallelization 2017-10-03
+	// Improve the retry loop, newer methodology
+	// Is rate limiting handled correctly?
+
+	// Increase attempt by a multiplier of the number of availability aones
+	// to account for the retries removed from the providers when the provisioner
+	// was parallelized.
+	task.azMachinesMutex.RLock()
+	var attemptsLeft int
+	if len(task.availabilityZoneMachines) > 0 {
+		attemptsLeft = task.retryStartInstanceStrategy.retryCount * len(task.availabilityZoneMachines)
+	} else {
+		// Some providers don't have availability zones
+		attemptsLeft = task.retryStartInstanceStrategy.retryCount
+	}
+	task.azMachinesMutex.RUnlock()
+
+	var result *environs.StartInstanceResult
+	for ; attemptsLeft >= 0; attemptsLeft-- {
 		attemptResult, err := task.broker.StartInstance(startInstanceParams)
 		if err == nil {
 			result = attemptResult
 			break
 		} else if attemptsLeft <= 0 {
 			// Set the state to error, so the machine will be skipped
-			// next time until the error is resolved, but don't return
-			// an error; just keep going with the other machines.
+			// next time until the error is resolved.
```

Comment:
If this happens, I think you should remove the machine from the AZ map? It's going to return nil (unless the controller is broken), but the machine hasn't been provisioned.

Reply:
if either setupToStartMachine() or startMachine() return an error, the machine is removed from the AZ map up in the goroutine.

Reply:
indeed, but as I said, this will return a *nil* error value. this returns the result of setErrorStatus, which will typically be nil; it'll only be non-nil if the status couldn't be updated

Reply:
setErrorStatus() returns an error now.  having it return nil was causing problems, but it looks like a different solution is needed.

Reply:
I'd like to know what problems it's causing. Having setErrorStatus return an error will cause other problems, namely when people like me expect it to return nil, and are surprised to find it doesn't.